Fix control variable initialization

SoilRos opened the pull request:
### Description 
Fixes initialization order bug stated in https://github.com/oneapi-src/oneTBB/issues/1341#issuecomment-2054095330

Fixes #1341

- [x] - git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/oneapi-src/oneTBB/blob/master/CONTRIBUTING.md#pull-requests) for details)_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [x] bug fix - _change that fixes an issue_
- [ ] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [ ] added - _required for new features and some bug fixes_
- [x] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [ ] needs to be updated
- [x] not needed

### Breaks backward compatibility
- [ ] Yes
- [x] No
- [ ] Unknown

### Notify the following users
@pavelkumbrasev

### Other information

kboyarinov commented on src/tbb/global_control.cpp:
```diff
@@ -139,13 +141,20 @@ class alignas(max_nfs_size) lifetime_control : public control_storage {
-static terminate_on_exception_control terminate_on_exception_ctl;
-static lifetime_control lifetime_ctl;
-static control_storage *controls[] = {&allowed_parallelism_ctl, &stack_size_ctl, &terminate_on_exception_ctl, &lifetime_ctl};
+static std::array<std::unique_ptr<control_storage>, 4> controls;
+static std::once_flag control_storage_flag;
```
We have discussed the solution briefly within the team and it seems like it can result in more undefined behavior since `std::unique_ptr` and `std::once_flag` are not trivial structures and it can be unsafe to access them before initialization (in particular if `global_control_init()` would be called before constructing the `controls` and `control_storage_flag`.

We propose to have global static pointer to the storage containing all required global control variables and dynamically allocate it inside `governor::acquire_resources()` (which would be called during the first access to TBB) and deallocate inside `governor::release_resources ()`.
@pavelkumbrasev, feel free to add if I missed something.

What do you think about the proposed change?

SoilRos replied:
I am _very_ unfamiliar with TBB internal code. I suspected that this solution may introduce more initialization problems as you pointed out, but I did not know where TBB initialized and destroyed things... so `unique_ptr` was the "easy" answer. With your hint we can give it a try on the `governor::[acquire|release]_resources` functions.

SoilRos commented:
Previous commit fails due to C++14 compatibility on `new` alignment. Now this should hopefully be fixed.

phprus commented on src/tbb/global_control.cpp:
```diff
@@ -149,17 +152,20 @@ static control_storage_instance* control_storage_instance_ptr = nullptr;
 
 void global_control_acquire() {
-    control_storage_instance_ptr = new control_storage_instance{};
+    auto ptr = aligned_alloc(max_nfs_size, sizeof(control_storage_instance));
```
MSVC does not support ``aligned_alloc``/``free``. Use ``_aligned_malloc`` and ``_aligned_free`` instead.

kboyarinov replied:
I suggest to use special functions: `cache_aligned_allocate()` and `cache_aligned_deallocate()`

pavelkumbrasev commented on src/tbb/global_control.cpp:
```diff
@@ -139,11 +142,31 @@ class alignas(max_nfs_size) lifetime_control : public control_storage {
+        ptr = nullptr;
+    }
+    control_storage_instance_ptr->~control_storage_instance();
+    free(std::exchange(control_storage_instance_ptr, nullptr));
```
`std::exchange` is not available in C++11

pavelkumbrasev commented on src/tbb/global_control.cpp:
```diff
@@ -139,11 +142,31 @@ class alignas(max_nfs_size) lifetime_control : public control_storage {
+    lifetime_control                lifetime_ctl;
+};
+static control_storage_instance* control_storage_instance_ptr = nullptr;
+static control_storage *controls[4] = {nullptr, nullptr, nullptr, nullptr};
```
Do we really need an extra variable + extra class? Can we just use pointer to array?

SoilRos replied:
It’s not really needed, I was trying to follow @kboyarinov’s instructions but I guess I understood them wrongly. So you meant that each control variable is allocated separately, right?

One argument for the struct is that it only needs one allocation so data is more compact.   With individual allocations you have to hard-code the derived class twice too anyways (construction/destruction). On the other hand, one can argue that the struct is slightly less maintainable, so I’d understand if you still prefer one allocation per class.

SoilRos replied:
Nvm, I just realized that every class sits in it’s own cache line so the “compact” argument doesn’t really hold…

pavelkumbrasev commented on src/tbb/global_control.cpp:
```diff
@@ -27,6 +28,7 @@
 #include "misc.h"
 
 #include <atomic>
+#include <new>
```
Do we really need this?

pavelkumbrasev commented on src/tbb/global_control.cpp:
```diff
@@ -139,11 +141,22 @@ class alignas(max_nfs_size) lifetime_control : public control_storage {
-static terminate_on_exception_control terminate_on_exception_ctl;
-static lifetime_control lifetime_ctl;
-static control_storage *controls[] = {&allowed_parallelism_ctl, &stack_size_ctl, &terminate_on_exception_ctl, &lifetime_ctl};
+static control_storage *controls[4] = {nullptr, nullptr, nullptr, nullptr};
```
```suggestion
static control_storage* controls[] = {nullptr, nullptr, nullptr, nullptr};
```

kboyarinov commented on src/tbb/global_control.cpp:
```diff
@@ -17,6 +17,7 @@
 #include "oneapi/tbb/detail/_config.h"
```
Please update the last copyright year to 2024 for both files to satisfy the copyright check.

pavelkumbrasev reviewed (commented):
Overall the patch looks good!

kboyarinov reviewed (commented):
Excluding the copyright comment - LGTM

SoilRos commented:
I addressed your reviews and squashed the commits, Thanks!

I was thinking that we may add a test for this with a global observer as presented in the original issue #1341. I just do not know where to put it. Do you have a hint where it can go? Or do you think it's not necessary?

pavelkumbrasev commented:
> I was thinking that we may add a test for this with a global observer as presented in the original issue #1341. I just do not know where to put it. Do you have a hint where it can go? Or do you think it's not necessary?

As I understand we cannot reproduce this failure without compiling TBB statically.  I don't know if we previously added separate tests for a specific platform.

SoilRos commented:
Alright, then this is ready from my side. I just amended my commit with the copyright comment (I hope it's correct). There is no changes in the code since the last CI run.

pavelkumbrasev commented:
> Maybe add tests for static build?

It might make sense since WASM compile TBB statically. Do you want to contribute it ? 😄

kboyarinov reviewed (approved):
LGTM

phprus commented:
> > Maybe add tests for static build?
> 
> It might make sense since WASM compile TBB statically. Do you want to contribute it ? 😄

PR: #1350
